### PR TITLE
Gpkg project storage

### PR DESCRIPTION
**Pull request by elpaso**
Still needs some polishing but it works

qgis_project_gpkg_store

# TODO

- [x] metadata username and change time

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -5643,16 +5643,20 @@ void QgisApp::fileOpenAfterLaunch()
...
   }
 
-  if ( !projPath.endsWith( QLatin1String( ".qgs" ), Qt::CaseInsensitive ) &&
-       !projPath.endsWith( QLatin1String( ".qgz" ), Qt::CaseInsensitive ) )
+  // Is this a storage based project?
+  const bool projectIsFromStorage { QgsApplication::instance()->projectStorageRegistry()->projectStorageFromUri( projPath )  };
+
+  if ( !( projectIsFromStorage ||
+          projPath.endsWith( QLatin1String( ".qgs" ), Qt::CaseInsensitive ) ||
+          projPath.endsWith( QLatin1String( ".qgz" ), Qt::CaseInsensitive ) ) )
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp]: A particular reason to change from `!a && !b && !c` to `!( a || b || c )`? Maybe it's just me, but I find the first easier to read.

File: src/providers/ogr/qgsgeopackageprojectstoragedialog.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
+/***************************************************************************
+    qgsgeopackageprojectstoragedialog.cpp
+    ---------------------
+    begin                : April 2018
+    copyright            : (C) 2018 by Martin Dobias
```
Comment (m-kuhn) [inline src/providers/ogr/qgsgeopackageprojectstoragedialog.cpp]: Leftover?

**Comment by m-kuhn** [comment]
Nice one!

File: src/app/qgisapp.cpp (hunk 1 above)
Comment (elpaso) [reply src/app/qgisapp.cpp]: Matter of tastes I guess.

Comment (m-kuhn) [reply src/app/qgisapp.cpp]: Not sure on that... Sequential parsing is probably easier than than inverting a grouped operation. Or something along these lines.
Anyway, not gonna stop this nice pull request on this one.

**Comment by Gustry** [comment]
Very nice addition!

Just a question, according to your GIF, does it work only for GPKG which have a "connection" under the "geopackage" tab in the datasource manager?

**Comment by Gustry** [comment]
Another question related to https://github.com/qgis/QGIS/pull/9381/commits/5d404cbd53628fe3d1055b310579ced4ffb5a8f9 can we see the list of layers/groups without opening the project? CF http://changelog.qgis.org/en/qgis/version/3.6.0/#explore-qgsqgz-project-files-browser

**Comment by elpaso** [comment]
> Very nice addition!
> 
> Just a question, according to your GIF, does it work only for GPKG which have a "connection" under the "geopackage" tab in the datasource manager?

Yes: that is the same behavior of the postgresql implementation.

**Comment by elpaso** [comment]
> Another question related to [5d404cb](https://github.com/qgis/QGIS/commit/5d404cbd53628fe3d1055b310579ced4ffb5a8f9) can we see the list of layers/groups without opening the project? CF http://changelog.qgis.org/en/qgis/version/3.6.0/#explore-qgsqgz-project-files-browser

Not for now, same implementation of the postgresql implementaion, it would be nice to add though.

**Comment by nyalldawson** [comment]
> Not for now, same implementation of the postgresql implementation, it would be nice to add though.

Another thing we should consider high on the future wishlist would be to show embedded project items in the browser from postgis/gpkg sources. (I think this would actually give you @gustry's request for "free").

**Comment by wonder-sk** [comment]
@elpaso great stuff... a small feature request from me :) can we make the "connection" combo box editable and make it accept gpkg filenames instead of just allowing stored connections? (+ a button to browse for gpkg files) Personally when I work with geopackage files, I never create connections and just load them directly from file system...

**Comment by nicolasmcfadden** [comment]
I'm sorry if this is not the right place to comment on this but I tried this great new feature and get the following error:

image

**Comment by elpaso** [comment]
@nicolasmcfadden you should file a bug report on https://issues.qgis.org/issues/
Please make sure to include a step-by-step procedure that allows to reproduce the issue on the developer machine.

**Pull request merged**